ssl/priority_queue.c

```diff
@@ -0,0 +1,390 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include "internal/priority_queue.h"
+#include "internal/safe_math.h"
+
+OSSL_SAFE_MATH_UNSIGNED(size_t, size_t)
+
+/* Fundamental operations:
+ *                        Binary Heap   Fibonacci Heap
+ *  Get smallest            O(1)          O(1)
+ *  Delete any              O(log n)      O(log n) average but worst O(n)
+ *  Insert                  O(log n)      O(1)
+ *
+ * Not supported:
+ *  Merge two structures    O(log n)      O(1)
+ *  Decrease key            O(log n)      O(1)
+ *  Increase key            O(log n)      ?
+ *
+ * The Fibonacci heap is quite a bit more complicated to implement and has
+ * larger overhead in practice.  We favour the binary heap here.  A multi-way
+ * (ternary or quaternary) heap might elicit a performance advantage via better
+ * cache access patterns.
+ */
+
+struct pq_heap_s {
+    void *data;     /* User supplied data pointer */
+    size_t index;   /* Constant index in elements[] */
+};
+
+struct pq_elem_s {
+    size_t posn;    /* Current index in heap[] or link in free list */
+#ifndef NDEBUG
+    int used;       /* Debug flag indicating that this is in use */
+#endif
+};
+
+struct ossl_pqueue_st
+{
+    struct pq_heap_s *heap;
+    struct pq_elem_s *elements;
+    int (*compare)(const void *, const void *);
+    size_t htop;        /* Highest used heap element */
+    size_t hmax;        /* Allocated heap & element space */
+    size_t freelist;    /* Index into elements[], start of free element list */
+};
+
+/*
+ * The initial and maximum number of elements in the heap.
+ */
+static const size_t min_nodes = 8;
+static const size_t max_nodes =
+        SIZE_MAX / (sizeof(struct pq_heap_s) > sizeof(struct pq_elem_s)
+                    ? sizeof(struct pq_heap_s) : sizeof(struct pq_elem_s));
+
+#ifndef NDEBUG
+/* Some basic sanity checking of the data structure */
+# define ASSERT_USED(pq, idx)                                               \
+    OPENSSL_assert(pq->elements[pq->heap[idx].index].used);                 \
+    OPENSSL_assert(pq->elements[pq->heap[idx].index].posn == idx)
+# define ASSERT_ELEM_USED(pq, elem)                                         \
+    OPENSSL_assert(pq->elements[elem].used)
+#else
+# define ASSERT_USED(pq, idx)
+# define ASSERT_ELEM_USED(pq, elem)
+#endif
+
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth factor is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use an expansion factor of 8 / 5 = 1.6
+ */
+static ossl_inline int compute_pqueue_growth(size_t target, size_t current)
+{
+    int err = 0;
+
+    while (current < target) {
+        if (current >= max_nodes)
```

Comment:
Maybe more this if above the loop and just return after line 98?

Reply:
No, this is detecting a hard overflow on the size.

Multiple expansions are not possible currently, but if they ever are, this needs to be here.

q.v. the stack code.